src/toil/fileStores/__init__.py

```diff
@@ -23,6 +25,33 @@ def make_public_dir(dirName: str) -> str:
     return dirName
 
 
+def unused_dir_path(prefix: str) -> str:
+    """
+    Try to make a random directory name with length 4 that doesn't exist, with the given prefix.
+
+    Otherwise, try length 5, length 6, etc, up to a max of 32 (len of uuid4 with dashes replaced).
+
+    This function's purpose is mostly to avoid having long file names when generating directories.
+
+    If somehow this fails, which should be incredibly unlikely, default to a normal uuid4, which was
+    our old default.
+    """
+    for i in range(4, 32 + 1):  # make random uuids and truncate to lengths starting at 4 and working up to max 32
+        for _ in range(10):  # make 10 attempts for each length
+            truncated_uuid = str(uuid4()).replace('-', '')[:i]
+            generated_dir_path = os.path.join(prefix, truncated_uuid)
+            if not os.path.exists(generated_dir_path):
+                return generated_dir_path
+    this_should_never_happen = os.path.join(prefix, str(uuid4()))
+    assert not os.path.exists(this_should_never_happen)
+    return this_should_never_happen
+
+
+def make_unique_public_dir(prefix: str) -> str:
```

Comment:
Are multiple processes/threads ever able to run this at the same time?

When we use UUIDs for making directories, we don't have to worry about races, because UUID collisions are effectively impossible due to their length. When our directory names might be as short as 5 characters, collisions are very possible. This would probably manifest as the mkdir in make_public_dir failing because the directory already exists.

[`tempfile.mkdtemp`](https://docs.python.org/3/library/tempfile.html#tempfile.mkdtemp) is the Right Way to make a unique subdirectory. If for some reason we can't do that, I would unify the name-generation and directory-creation into one loop, using a failed mkdir as the test for the directory already existing.

Were we using UUIDs to avoid collisions not just on this machine, but also on other machines?

Reply:
You're right, I should have accounted for races.  I think the try/except on a `mkdir` makes sense to me.

`tempfile.mkdtemp` creates a ` directory [that] is readable, writable, and searchable only by the creating user ID.` and while I'm unsure that that's a case we have to account for naturally, I suspect that this might make it difficult to inspect directories by an outside user in the event of a failed workflow trying to go through the files so I'd rather avoid it.